Unix/http/httpclientauth.c

```diff
@@ -211,13 +217,224 @@ typedef struct _Gss_Extensions
+       };
+
+  static Probable_Cause_Data AUTH_ERROR_CRED_FILE = {
+              ERROR_ACCESS_DENIED,
+              WSMAN_CIMERROR_PROBABLE_CAUSE_AUTHENTICATION_FAILURE,
+              MI_T("credentials file and directory does not exist") 
+       };
+
+  static Probable_Cause_Data AUTH_ERROR_CRED_DIR_PERMS = {
+              ERROR_ACCESS_DENIED,
+              WSMAN_CIMERROR_PROBABLE_CAUSE_AUTHENTICATION_FAILURE,
+              MI_T("credentials file directory permissions must be user-only") 
+       };
+
+  static Probable_Cause_Data AUTH_ERROR_CRED_FILE_PERMS = {
+              ERROR_ACCESS_DENIED,
+              WSMAN_CIMERROR_PROBABLE_CAUSE_AUTHENTICATION_FAILURE,
+              MI_T("credentials file permissions must be user-only") 
+       };
+
+
+    if (ntlm_user_file)
+    {
+        // Check the permissions on the directory. They need to be read-only owner
+        
+        {
+            struct stat buf = {0};
+            char *ntlm_cred_dir = strdup(ntlm_user_file);
+            char* p = strrchr(ntlm_cred_dir, '/');
+            if (p)
+            {
+                *p = '\0';
+            }
+
+            int rtn = stat(ntlm_cred_dir, &buf);
+            if (rtn < 0)
+            {
+                free(ntlm_cred_dir);
+                _g_gssClientState.probableCause = &AUTH_ERROR_CRED_FILE;
+                goto Err1;
+               
+            }
+
+            if (!S_ISDIR(buf.st_mode))
+            {
+               // Not a directory? Unlikely, but complain and issue error
+                free(ntlm_cred_dir);
+                _g_gssClientState.probableCause = &AUTH_ERROR_CRED_FILE;
+                goto Err1;
+            }
+
+            // Acceptable dir will be user only, no others
+            if (!(buf.st_mode & S_IRUSR) || ( buf.st_mode & (S_IRWXG|S_IRWXO)))
+            {
+                free(ntlm_cred_dir);
+                _g_gssClientState.probableCause = &AUTH_ERROR_CRED_DIR_PERMS;
+                goto Err1;
+            }
+            free(ntlm_cred_dir);
+        }
+            
+        // Check the file for the same permissions
+        {
+            struct stat buf = {0};
+            int rtn = stat(ntlm_user_file, &buf);
+
+            if (rtn < 0)
+            {
+               // No directory? Complain and issue error
+                _g_gssClientState.probableCause = &AUTH_ERROR_CRED_DIR;
+                goto Err;
+            }
+
+            if (S_ISDIR(buf.st_mode) || S_ISLNK(buf.st_mode))
+            {
+               // Not a file? Unlikely, but complain and issue error
+                _g_gssClientState.probableCause = &AUTH_ERROR_CRED_FILE;
+                goto Err;
+            }
+
+            // Acceptable dir will be user only, no others
+            if (!(buf.st_mode & S_IRUSR) || ( buf.st_mode & (S_IRWXG|S_IRWXO)))
+            {
+                _g_gssClientState.probableCause = &AUTH_ERROR_CRED_DIR_PERMS;
+                goto Err;
+            }
+        }
+
+    }
+    else 
```

Comment:
why duplicate most of the code. Why not at the beginning say if env is null make the path, then do most of the same checking once rather than twice.